agent/engine/containermetadata/metadata_manager.go

```diff
@@ -0,0 +1,121 @@
+// Copyright 2017 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License"). You may
+// not use this file except in compliance with the License. A copy of the
+// License is located at
+//
+//	http://aws.amazon.com/apache2.0/
+//
+// or in the "license" file accompanying this file. This file is distributed
+// on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+// express or implied. See the License for the specific language governing
+// permissions and limitations under the License.
+
+package containermetadata
+
+import (
+	"fmt"
+	"os"
+	"time"
+
+	"github.com/aws/amazon-ecs-agent/agent/api"
+	"github.com/aws/amazon-ecs-agent/agent/config"
+	"github.com/cihub/seelog"
+)
+
+const (
+	inspectContainerTimeout = 30 * time.Second
+)
+
+// MetadataManager is an interface that allows us to abstract away the metadata
+// operations
+type MetadataManager interface {
+	SetContainerInstanceArn(string)
+	CreateMetadata([]string, *api.Task, *api.Container) ([]string, error)
+	UpdateMetadata(string, *api.Task, *api.Container) error
+	CleanTaskMetadata(*api.Task) error
+}
+
+// metadataManager implements the MetadataManager interface
+type metadataManager struct {
+	client               dockerDummyClient
+	cfg                  *config.Config
+	containerInstanceArn string
+}
+
+// NewMetadataManager creates a metadataManager for a given DockerTaskEngine settings.
+func NewMetadataManager(client dockerDummyClient, cfg *config.Config) MetadataManager {
+	seelog.Debug("Creating metadata manager from configuration: %s", cfg)
+	return &metadataManager{
+		client: client,
+		cfg:    cfg,
+	}
+}
+
+// SetContainerInstanceArn sets the metadataManager's ContainerInstanceArn which is not available
```

Comment:
Yeah, I think what you have here is reasonable given the state of the agent right now.  What I was pointing out is that the behavior in the agent today isn't super reasonable; it would probably be better if the task engine was only constructed *after* we had successfully registered.  I'm not asking you to fix that.